python/ray/serve/_private/deployment_scheduler.py

```diff
@@ -311,8 +467,75 @@ def schedule(
+        target_labels: Dict[str, Any] = None,
+        target_node_id: str = None,
+    ):
+        """Schedule a replica from a scheduling request."""
+
+        deployment_id = scheduling_request.deployment_id
+        replica_name = scheduling_request.replica_name
+        placement_group = None
+
+        scheduling_strategy = default_scheduling_strategy
+        if scheduling_request.placement_group_bundles is not None:
+            placement_group_strategy = (
+                scheduling_request.placement_group_strategy
+                if scheduling_request.placement_group_strategy
+                else "PACK"
+            )
+            pg = self._create_placement_group_fn(
+                scheduling_request.placement_group_bundles,
+                placement_group_strategy,
+                _soft_target_node_id=target_node_id,
+                lifetime="detached",
+                name=scheduling_request.actor_options["name"],
+            )
+            scheduling_strategy = PlacementGroupSchedulingStrategy(
+                placement_group=pg,
+                placement_group_capture_child_tasks=True,
+            )
+        elif target_labels is not None:
+            scheduling_strategy = NodeLabelSchedulingStrategy(
+                hard={}, soft=target_labels
+            )
+        elif target_node_id is not None:
+            scheduling_strategy = NodeAffinitySchedulingStrategy(
+                node_id=target_node_id, soft=True, _spill_on_unavailable=True
+            )
+
+        actor_options = copy.copy(scheduling_request.actor_options)
+        if scheduling_request.max_replicas_per_node is not None:
+            if "resources" not in actor_options:
+                actor_options["resources"] = {}
+            # Using implicit resource (resources that every node
+            # implicitly has and total is 1)
+            # to limit the number of replicas on a single node.
+            actor_options["resources"][
+                f"{ray._raylet.IMPLICIT_RESOURCE_PREFIX}"
+                f"{deployment_id.app_name}:{deployment_id.name}"
+            ] = (1.0 / scheduling_request.max_replicas_per_node)
+
+        actor_handle = scheduling_request.actor_def.options(
+            scheduling_strategy=scheduling_strategy,
+            **actor_options,
+        ).remote(*scheduling_request.actor_init_args)
+
+        del self._pending_replicas[deployment_id][replica_name]
+        self._on_replica_launching(
+            deployment_id, replica_name, target_node_id=target_node_id
+        )
+
+        if isinstance(scheduling_strategy, PlacementGroupSchedulingStrategy):
+            placement_group = scheduling_strategy.placement_group
+
+        scheduling_request.on_scheduled(actor_handle, placement_group=placement_group)
+
     @abstractmethod
-    def detect_compact_opportunities(self) -> Tuple[str, float]:
+    def detect_compact_opportunities(self) -> Tuple[Optional[str], Optional[float]]:
```

Comment:
should it be `Optional[Tuple[str, float]]`

Reply:
Will address in https://github.com/ray-project/ray/pull/43755